extension/cached_registry_test.go

```diff
@@ -0,0 +1,111 @@
+package extension
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/google/trillian/storage"
+)
+
+func TestGetLogStorage(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	registry := NewMockRegistry(ctrl)
+	cachedRegistry := NewCachedRegistry(registry)
+
+	ls1 := storage.NewMockLogStorage(ctrl)
+	ls2 := storage.NewMockLogStorage(ctrl)
+	registry.EXPECT().GetLogStorage(int64(1)).Times(1).Return(ls1, nil)
+	registry.EXPECT().GetLogStorage(int64(2)).Times(1).Return(ls2, nil)
+
+	var tests = []struct {
+		treeID int64
+		want   storage.LogStorage
+	}{
+		{1, ls1},
+		{1, ls1}, // Same key twice to test caching
+		{2, ls2},
+	}
+	for _, test := range tests {
+		got, err := cachedRegistry.GetLogStorage(test.treeID)
+		if err != nil {
+			t.Errorf("GetLogStorage(%v) failed with error: %v", test.treeID, err)
+		} else if got != test.want {
+			t.Errorf("GetLogStorage(%v) failed, want %q, got %q", test.treeID, test.want, got)
+		}
+	}
+}
+
+func TestGetLogStorageError(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	registry := NewMockRegistry(ctrl)
+	cachedRegistry := NewCachedRegistry(registry)
+
+	want := fmt.Errorf("Error getting log storage")
+	registry.EXPECT().GetLogStorage(int64(1)).Times(2).Return(nil, want)
+
+	// Run twice to make sure caching isn't doing anything funky
+	for i := 0; i < 2; i++ {
+		ls, err := cachedRegistry.GetLogStorage(1)
+		if err == nil || err != want {
```

Comment:
I guess you don't need the == nil bit?
Also, it's apparently idiomatic in Go to use the general switch rather than if/else if/... chains:
```go
switch {
case err != want:
   t.Error...
case ls != nil:
   t.Error...
}
```